File: LEVEL2/조이스틱.java
class Solution {
    public int solution(String name) {
        int answer = 0;
        int res = name.length() - 1;
        
        for(int i = 0; i < name.length(); i++) {
            answer += Math.min(name.charAt(i) - 'A', 'Z' - name.charAt(i) + 1);
            int next = i + 1;
        
            while(next < name.length() && name.charAt(next) == 'A') {
                next++;
            }

            res = Math.min(res, i + name.length() - next + i);
        }
        
        answer += res;
        
        return answer;
    }
}

File: LEVEL2/조이스틱.cpp
#include<string>
using namespace std;

int solution(string name) {
    int answer = 0;
    int p = 0;
    string cmp;
    
    for(int i = 0; i < name.size(); i++) {
        cmp += "A";
    }
    
    while(cmp != name) {
        int next = 0;
        int left = 0;
        int right = 0;
        
        for(int i = 0; i < name.size(); i++) {
            if(i + p < name.size()) {
                right = i + p;
            }
            
            else {
                right = i + p - name.size();
            }
            
            if(i <= p) {
                left = p - i;
            }
            
            else {
                left = p - i + name.size();
            }
            
            if(cmp[right] != name[right]) {
                next = right;
            }
            
            else if(cmp[left] != name[left]) {
                next = left;
            }
            
            else {
                continue;
            }
            
            answer += i;
            answer += min(name[next] - 'A', 'Z' - name[next] + 1);
            cmp[next] = name[next];
            
            break;
        }
        
        p = next;
    }
    
    return answer;
}